python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py

```diff
@@ -674,4 +809,27 @@ async def assign_replica(
         request, so it's up to the caller to time out or cancel the request.
         """
         replica = await self.choose_replica_for_query(query)
-        return replica.send_query(query), replica.replica_id
+        replica_id = replica.replica_id
+
+        # If the queue len cache is disabled or we're sending a request to Java,
+        # then directly send the query and hand the response back. The replica will
+        # never reject requests in this code path.
+        if not self._use_replica_queue_len_cache or replica.is_cross_language:
+            return replica.send_query(query, with_rejection=False), replica_id
+
+        while True:
+            obj_ref_gen = replica.send_query(query, with_rejection=True)
+            accepted = await self._handle_system_response(replica_id, obj_ref_gen)
+            if accepted:
+                if query.metadata.is_streaming:
+                    return obj_ref_gen, replica_id
+                else:
+                    # For non-streaming requests, resolve the generator to its next
+                    # object ref, which will contain the unary response.
+                    return await obj_ref_gen.__anext__(), replica_id
+
+            # If the replica rejects the request, retry the scheduling process.
+            # Place the request on the front of the queue to avoid tail latencies.
+            # TODO(edoakes): this retry procedure is not perfect because it'll reset the
+            # process of choosing candidates replicas (i.e., for locality-awareness).
+            replica = await self.choose_replica_for_query(query, emplace_front=True)
```

Comment:
This is a good point. A priority queue sorted by timestamp seems to be ideally what we want. Let me take this as a follow-up and benchmark/profile it to see if the overhead is significant (I doubt it would be).

Reply:
Actually I just went ahead and implemented this by continuing to use the `deque`.

Because retries should be rare and only happen under contention, it's ok for this path to be a bit slower, so what I've done is iterated over the queue and inserted it in the correct place on this path. See the `is_retry` logic in `choose_replica_for_request`.